kale/embed/mpca.py

```diff
@@ -1,181 +1,324 @@
+                X_transformed = X_transformed[:, self.idx_order]
+                X_transformed = X_transformed[:, :n_components]
 
+        return X_transformed
 
-def MPCA_(X, variance_explained=0.97, max_iter=1):
-    """
-    Parameter:
-        X: array-like, ndarray of shape (I1, I2, ..., n_samples)
-        variance_explained: ration of variance to keep (between 0 and 1)
-        max_iter: max number of iteration
-    ----------
-    Return:
-        tPs: list of projection matrices
-    """
-    dim_in = X.shape
-    n_spl = dim_in[-1]
-    n_dim = X.ndim
-    # Is = dim_in[:-1]
-    Xmean = np.mean(X, axis=-1)
-    X_ = np.zeros(X.shape)
-    # init
-    Phi = dict()
-    Us = dict()  # eigenvectors
-    Vs = dict()  # eigenvalues
-    cums = dict()  # cumulative distribution of eigenvalues
-    tPs = []
-    for i in range(n_spl):
-        X_[..., i] = X[..., i] - Xmean
-        for j in range(n_dim-1):
-            X_i = unfold(X_[..., i], mode=j)
-            if j not in Phi:
-                Phi[j] = 0
-            Phi[j] = Phi[j] + np.dot(X_i, X_i.T)
-
-    for i in range(n_dim-1):
-        eig_vals, eig_vecs = np.linalg.eig(Phi[i])
-        idx_sorted = eig_vals.argsort()[::-1]
-        Us[i] = eig_vecs[:, idx_sorted]
-        Vs[i] = eig_vals[idx_sorted]
-        sum_ = np.sum(Vs[i])
-        for j in range(Vs[i].shape[0] - 1, 0, -1):
-            if np.sum(Vs[i][j:]) / sum_ > (1 - variance_explained):
-                cums[i] = j + 1
-                break
-        tPs.append(Us[i][:, :cums[i]].T)
-
-    for i_iter in range(max_iter):
-        Phi = dict()
-        for i in range(n_dim-1):
-            # dim_in_ = dim_in[i]
-            if i not in Phi:
-                Phi[i] = 0
-            for j in range(n_spl):
-                X_i = X_[..., j]
-                Xi_ = multi_mode_dot(X_i, [tPs[m] for m in range(n_dim-1) if m != i],
-                                     modes=[m for m in range(n_dim) if m != i])
-                tXi = unfold(Xi_, i)
-                Phi[i] = np.dot(tXi, tXi.T) + Phi[i]
+    def inverse_transform(self, X):
+        """Transform data in the shape of reduced dimension 
+        back to the original shape
 
-            eig_vals, eig_vecs = np.linalg.eig(Phi[i])
-            idx_sorted = eig_vals.argsort()[::-1]
-            tPs[i] = eig_vecs[:, idx_sorted]
-            tPs[i] = tPs[i][:, :cums[i]].T
+        Args:
+            X (ndarray): New data, shape (P_1, P_2, ..., P_N, n_samples), where P_1, P_2, ..., P_N 
+                are the reduced dimensions of of corresponding mode (1, 2, ..., N), respectively.
 
-    return tPs
+        Returns:
+            ndarray: Data in original shape, shape (I_1, I_2, ..., I_N, n_samples)
+        """
+        _check_dim_shape(X, self.ndim, self.shape_out)
+        return multi_mode_dot(X, self.proj_mats, modes=[m for m in range(self.ndim - 1)], transpose=True)
 
 
+# The following implementation kept for a tensorly function
```

Comment:
What's the purpose of keeping this? And it is not synchronised with recent update.

Reply:
will remove in the next update